lib/event/manager.go

```diff
@@ -137,44 +141,44 @@ func (e *Manager) Start(renderer interfaces.Renderer) {
 					e.log.Error(err.Error())
 				}
 
-				e.mu.Lock()
-
 				// Iterate listeners
-				for _, listener := range e.listeners[event.Name] {
-
-					if !listener.expired {
-						// Call listener, perhaps with data
-						if event.Data == nil {
-							go listener.callback()
-						} else {
-							unpacked := event.Data.([]interface{})
-							go listener.callback(unpacked...)
-						}
+				for currentIndex, listener := range e.listeners[event.Name] {
+
+					// Call listener, perhaps with data
+					if event.Data == nil {
+						go listener.callback()
+					} else {
+						unpacked := event.Data.([]interface{})
+						go listener.callback(unpacked...)
 					}
 
-					// Update listen counter
+					// Decrement counter if its a non-persistent listener
 					if listener.counter > 0 {
-						listener.counter = listener.counter - 1
+						// fields about to change; enter Mutex
+						e.mu.Lock()
+						listener.counter--
+
+						// expiration condition == counter WAS 1 but is NOW 0
 						if listener.counter == 0 {
-							listener.expired = true
+							// this listener has just expired; remove it NOW
+							// see fast method: https://yourbasic.org/golang/delete-element-slice/
+							// https://play.golang.org/p/j0YjKUN0NL1
+							lastIndex := len(e.listeners[event.Name]) - 1
+
+							// overwrite expired currentIndex listener with lastIndex listener
+							e.listeners[event.Name][currentIndex] = e.listeners[event.Name][lastIndex]
+
+							// remove the (now cloned) lastIndex listener from slice
+							e.listeners[event.Name] = e.listeners[event.Name][:lastIndex]
```

Comment:
so the problem I found with this approach is that you are iterating on `e.listeners[event.Name]` at line 145 which will cause issues if you alter the struct. This may work if iterating on a copy of the struct ?

Reply:
This is pretty much the code I had in today but found race conditions if you smashed the `On` methods

Reply:
>...which will cause issues if you alter the struct. This may work if iterating on a copy of the struct ?

So @leaanthony you said `struct`, I think you meant `slice` (`e.listeners[event.Name]`)?  I'm not trying to be pedantic, but concurrency is a devil and I just want to be clear about what changes you are concerned about. I am **not** concerned about changes to other fields in the `struct`, but yes I am **very** concerned about changes in the `slice`. _Assuming you meant `slice`..._

I mentally see that race condition now; we've made an error here. When you are smashing `.On()` you would be increasing `len(slice)` in a Goroutine possibly at the same time when `.Start()` has already computed `lastIndex` for the clone and truncate operation in another Goroutine. When that happens then clone and truncate will not work properly. With a `lastIndex` value that is _no longer_ the last index, it will truncate the intended target **and** the new one just added to the end of the `slice`!

So the answer is to ensure that **all** code that changes the `slice` is wrapped in a `mutex`. I have code reviewed the file again, and other than `.Start()` where this PR originally made changes, the **only** other location where the `slice` is altered is `.addEventListener()`. All the other methods delegate the task of adding a new listener to the `slice` there. 

_I have just now added a new commit to this PR that does this, and with all locations now wrapped in a `mutex`, your race conditions go away_ 💯

Reply:
I've created a demo of the race condition [in the playground](https://play.golang.org/p/URLCzsi7gcp)

Reply:
Yeah i meant slice. I also wrapped it all in a mutex but that didn't stop the race condition. I'm not sure what the range semantics are and whether it's safe to remove items from a slice you are iterating over. Judging by online comments, it isn't. My first approach to this was to add all valid listeners to a new slice and reassing that but that's heavy. I think the best bet is to make a copy of the listener slice, iterate over that but update the original slice.

Reply:
In this scenraio, the copy approach would be best. https://freshman.tech/snippets/go/copy-slices/

Reply:
Where did we get to with this?